sys/cheri/cheri_usercap.c

```diff
@@ -156,3 +171,57 @@ _cheri_capability_build_user_rwx_unchecked(uint32_t perms, ptraddr_t basep,
 	return (cheri_setoffset(cheri_andperm(cheri_setbounds(
 	    cheri_setoffset(userspace_root_cap, basep), length), perms), off));
 }
+
+/*
+ * Try to store a tagged capability in *out, derived from an untagged
+ * "bag of bits" in in.  If a tagged capability cannot be derived,
+ * return false and leave *out unchanged.
+ */
+bool
+ptrace_derive_cap(struct thread *td, uintcap_t in,
+    uintcap_t *out)
+{
+	struct thread *td2;
+	void * __capability cap;
+	void * __capability sealcap;
+
+	/*
```

Comment:
It's debatable if trying the target thread first is a worthwhile optimization.  Otherwise this function could probably just take a proc pointer instead.